### PR TITLE
Add BoringSSL support for Windows

### PR DESCRIPTION
**Pull request by NikitaNikolaenko**

### CODE REVIEW

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:20]: If I'm reading this right, this makes this filegroup only visible to this package and its subfolders. But this package has no subfolders, meaning this is only visible to this package.

That sounds like private visibility: should we use that instead to be more clear? If so, that's the default visibility value: https://docs.bazel.build/versions/main/be/common-definitions.html#common.visibility so just remove this line.

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:31]: Similar to above, we can just remove the visibility line here since private is the default.

Here and elsewhere in this file.

File: bazel/deps.bzl
Comment (CodingCanuck) [inline bazel/deps.bzl:13]: Can you comment on why we use boringssl?

Reading https://boringssl.googlesource.com/boringssl I see: "Although BoringSSL is an open source project, it is not intended for general use, as OpenSSL is. We don't recommend that third parties depend upon it. Doing so is likely to be frustrating because there are no guarantees of API or ABI stability."

Which makes me think we should use something else. A comment explaining why we chose this (in spite of this warning) would be useful.

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel]: Lots of "whys" here that could use documenting: Why use a filegroup instead of a cc_library or similar?

Why are these libraries srcs, rather than deps?

Should each of the sub-libraries have its own cc_library target that this just groups together?

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel]: Please document: why?

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:44]: As discussed: would it make sense to just symlink these? If so, please symlink instead of copy. If not, please document why.

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:57]: Per above: why is this in srcs rather than deps? Please document.

Further: why do we need the filegroup here at all if we're eventually just exposing this through a cc_library? What happens if we just add the raw boringssl rules as deps to this rule, and give this rule no srcs?  (If this doesn't work, please document why)

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:13]: What did this line do, and why are we removing it? (Does removing this release build type switch us to a non-optimized build or something?)

Or ah, is this replaced by the select statement in the out_static_libs property of the libcurl_boringssl_windows target?

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:73]: While you're here, can you help to fix this comment? It ends partway through a sentence (but bazel will expect... what?)

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel]: Maybe pull out these defines into a LIBCURL_DEFINES variable or similar that can be shared across rules, so e.g.

LIBCURL_DEFINES = [ "CURL_STATICLIB" ]

cmake(name = "foo", defines = LIBCURL_DEFINES)

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel]: Should this have a TODO to not do this in order to make this build hermetic?

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:12]: Maybe call this curl_srcs or similar? "all" is a confusing name since this doesn't include things like boringssl.

File: BUILD.bazel
Comment (CodingCanuck) [inline BUILD.bazel:28]: optional: WDYT about moving all of this boringssl stuff into a separate BUILD.bazel file (e.g. a sub-package), so the concerns of building boringssl and building libcurl are separated?

**Review by CodingCanuck — APPROVED** [review]
Nice work getting this working! I know it was a struggle.

It doesn't look like this repo has any CI set up. So I'm not confident that this builds / works, because there's no evidence from a pre-commit automated build :)

Can you file an issue to set up CI for this stuff? (And aspirationally, do you think we can / should do that before submitting this PR?)

**Comment by NikitaNikolaenko** [comment]
> Nice work getting this working! I know it was a struggle.
> 
> It doesn't look like this repo has any CI set up. So I'm not confident that this builds / works, because there's no evidence from a pre-commit automated build :)
> 
> Can you file an issue to set up CI for this stuff? (And aspirationally, do you think we can / should do that before submitting this PR?)

We do use it in eventuals, see https://github.com/3rdparty/eventuals/pull/175

**Comment by CodingCanuck** [comment]
> We do use it in eventuals, see [3rdparty/eventuals#175](https://github.com/3rdparty/eventuals/pull/175)

Gotcha, so this is ~indirectly covered by another repo's CI?

I think it would still be good to set up CI for this repo so we block merges that would cause build breaks, rather than having those build breaks be discovered in downstream repos.

But that other PR's successful builds give me confidence for this PR :D (so CI setup doesn't need to block this PR)

**Comment by NikitaNikolaenko** [comment]
> > We do use it in eventuals, see [3rdparty/eventuals#175](https://github.com/3rdparty/eventuals/pull/175)
> 
> Gotcha, so this is ~indirectly covered by another repo's CI?
> 
> I think it would still be good to set up CI for this repo so we block merges that would cause build breaks, rather than having those build breaks be discovered in downstream repos.
> 
> But that other PR's successful builds give me confidence for this PR :D (so CI setup doesn't need to block this PR)

Yeah, I personally think that it's a great idea to implement a CI setup and even make some tests for it. Not sure if we have enough time for each and every repo :D

File: bazel/deps.bzl
Comment (benh) [reply bazel/deps.bzl:13]: This is a very fair point given the boringssl documentation, but given the current state of the industry the boringssl maintainers should probably update their documentation as there are tons of open source projects that depend on boringssl.

File: BUILD.bazel
Comment (benh) [reply BUILD.bazel:12]: @FolMing is following a pattern of the other `3rdparty/bazel-rules-*` repos here that have a `filegroup` named `all`. Consistency says keep the pattern, but acknowledged that this file is clearly more complicated. I like the suggestion of moving the `boringssl` build pieces into a `boringssl/BUILD.bazel` and just depending on targets in this file. Likewise I think adding more comments in that file around the why and how we make Windows happy by the magic of `filegroup`, `genrule`, and `cc_library` would be helpful for future debugging.

File: BUILD.bazel
Comment (NikitaNikolaenko) [reply BUILD.bazel:20]: Making it private breaks our eventuals build, so I used the most "private" option that I'm aware of.

File: BUILD.bazel
Comment (NikitaNikolaenko) [reply BUILD.bazel:31]: I guess we could remove visibility:private lines, BUT I've taken a habit of marking visibility as explicit as possible - it's easier for me to figure out what's public or what's private.

File: BUILD.bazel
Comment (NikitaNikolaenko) [reply BUILD.bazel:13]: Exactly!
I've been looking through cmake rule implementation and I've noticed that it checks whether our current build is a debug or an optimized one and appends the CMAKE_BUILD_TYPE by itself.
They could have documented it a little bit better though...

File: BUILD.bazel
Comment (benh) [reply BUILD.bazel:20]: So sounds like you know of a reason why you had to do something "special" ... then we need a comment that explains why!

File: BUILD.bazel
Comment (benh) [inline BUILD.bazel]: Just using "@boringssl//:crypto" and "@boringssl//:ssl" as our 'deps' as shown above gives doesn't work because the 'cmake' rule seems to try and force boringssl to create BOTH static and shared libraries the latter of which does not work on Windows. Our workaround is to create our own  "//:boringssl_lib" which only has the static libraries. For additional information see comments above.

**Review by benh — APPROVED** [review]

File: BUILD.bazel
Comment (NikitaNikolaenko) [reply BUILD.bazel:44]: Added a TODO, let's just copy for this PR.

File: BUILD.bazel
Comment (NikitaNikolaenko) [reply BUILD.bazel:73]: I think it's a viewing bug on your side! :D

File: BUILD.bazel
Comment (NikitaNikolaenko) [reply BUILD.bazel:28]: I personally think it's a good idea, but due to us having specific boringssl for libcurl let's not do that for now.

**Pull request merged**
